src/lib/Nbe.ml

```diff
@@ -784,11 +784,20 @@ struct
       let* var = top_var tp in
       m var
 
-
   let rec quote_con (tp : D.tp) con : S.t m =
     QuM.abort_if_inconsistent S.CofAbort @@
     match tp, con with
     | _, D.Abort -> ret S.CofAbort
+    | _, D.Cut {cut = (D.Var lvl, []); tp = TpDim} ->
+      (* for dimension variables, check to see if we can prove them to be
+         the same as 0 or 1 and return those instead if so. *)
+      let* eq0 = lift_cmp @@ CmpM.test_sequent [] (Cof.eq (D.DimVar lvl) D.Dim0) in
+      let* eq1 = lift_cmp @@ CmpM.test_sequent [] (Cof.eq (D.DimVar lvl) D.Dim1) in
+      if eq0 then ret S.Dim0
+      else if eq1 then ret S.Dim1
+      else
+        let+ ix = quote_var lvl in
+        S.Var ix
```

Comment:
Because these tests aren't lazy, it may be better to put the second test in the `else` branch.

Reply:
yeah, fair. i'll do that in a sec.

Reply:
just out of curiosity: how much of that is a hard learned knee-jerk reaction to avoid work when you can, versus actual knowledge about the cost of these tests? differently: in practice, how bad would you expect it to be to actually determine if a dimension variable is equal to 0 or 1?

Reply:
I have no idea the cost, so I would categorize it under "basic knee-jerk" --- generally speaking I don't expect dimension comparisons to be free, because it involves this persistent union-find structure (which is of course itself quite a bit less efficient than the standard imperative one).

Reply:
gotcha.